packages/govuk-frontend/src/govuk/components/header/header.mjs

```diff
@@ -52,22 +52,43 @@ export class Header extends GOVUKFrontendComponent {
     }
 
     this.$module = $module
-    this.$menuButton = $module.querySelector('.govuk-js-header-toggle')
-    this.$menu =
-      this.$menuButton &&
-      $module.querySelector(
-        `#${this.$menuButton.getAttribute('aria-controls')}`
-      )
-
-    if (
-      !(
-        this.$menuButton instanceof HTMLElement ||
-        this.$menu instanceof HTMLElement
-      )
-    ) {
+    const $menuButton = $module.querySelector('.govuk-js-header-toggle')
+
+    // Headers don't necessarily have a navigation.
+    // When they don't, the menu toggle won't be rendered
+    // by our macro (or may be omitted when writing plain HTML)
+    if (!$menuButton) {
       return this
     }
 
+    if (!($menuButton instanceof HTMLElement)) {
+      throw new ElementError($menuButton, {
+        componentName: 'Header',
+        identifier: '.govuk-js-header-toggle'
+      })
+    }
+
+    const menuId = $menuButton.getAttribute('aria-controls')
+    if (!menuId) {
+      throw new ElementError($menuButton, {
+        componentName: 'Header',
+        identifier: '$menuButton["aria-controls"]',
+        expectedType: 'string'
+      })
+    }
+
+    const $menu = document.getElementById(menuId)
+
+    if (!($menu instanceof HTMLElement)) {
+      throw new ElementError($menu, {
+        componentName: 'Header',
+        identifier: `#${menuId}`
+      })
+    }
+
+    this.$menu = $menu
+    this.$menuButton = $menuButton
```

Comment:
Thanks for doing this

Helps the compiler out a lot (narrows the types down) when adding to class fields _after_ the checks